### PR TITLE
Maximally share hashconsed subapplication nodes in Typeops.

### PR DESCRIPTION
**Pull request by ppedrot**
Probably way worse but just curious.

### CODE REVIEW

**Comment by ppedrot** [comment]
@coqbot bench

File: kernel/typeops.ml [hunk 1]
```diff
@@ -632,6 +632,19 @@ let push_rec_types (lna,typarray,_) env =
   let ctxt = Array.map2_i (fun i na t -> RelDecl.LocalAssum (na, lift i t)) lna typarray in
   Array.fold_left (fun e assum -> push_rel assum e) env ctxt
 
+let share_application tbl f args =
+  let rec share n =
+    if n <= 1 then None
```
Comment (SkySkimmer) [inline kernel/typeops.ml:637]: why <= instead of < ?

File: kernel/typeops.ml [hunk 2]
```diff
@@ -632,6 +632,19 @@ let push_rec_types (lna,typarray,_) env =
...
+    else
+      let sargs = Array.sub args 0 n in
+      let app = HConstr.of_kind_nohashcons (App (f, sargs)) in
```
Comment (SkySkimmer) [inline kernel/typeops.ml:640]: if we think the Array.sub and possibly the kind_to_constr in of_kind_hashcons are the main cost, we could have a lookup specialized for partial applications. ie something like
~~~diff
modified   kernel/hConstr.ml
@@ -172,6 +172,14 @@ let hash_annot = hash_annot Name.hash
 
 let hash_array hashf a = Array.fold_left (fun hash x -> Hashset.Combine.combine hash (hashf x)) 0 a
 
+let hash_app f args n =
+  let open Hashset.Combine in
+  let h = ref f.hash in
+  for i = 0 to n - 1 do
+    h := combine !h args.(i).hash
+  done;
+  combinesmall 7 !h
+
 let hash_kind = let open Hashset.Combine in function
   | Var i -> combinesmall 1 (Id.hash i)
   | Sort s -> combinesmall 2 (Sorts.hash s)
@@ -179,7 +187,7 @@ let hash_kind = let open Hashset.Combine in function
   | Prod (na,t,c) -> combinesmall 4 (combine3 (hash_annot na) t.hash c.hash)
   | Lambda (na,t,c) -> combinesmall 5 (combine3 (hash_annot na) t.hash c.hash)
   | LetIn (na,b,t,c) -> combinesmall 6 (combine4 (hash_annot na) b.hash t.hash c.hash)
-  | App (h,args) -> combinesmall 7 (Array.fold_left (fun hash c -> combine hash c.hash) h.hash args)
+  | App (h,args) -> hash_app h args (Array.length args)
   | Evar _ -> assert false
   | Const (c,u) -> combinesmall 9 (combine (Constant.SyntacticOrd.hash c) (UVars.Instance.hash u))
   | Ind (ind,u) -> combinesmall 10 (combine (Ind.SyntacticOrd.hash ind) (UVars.Instance.hash u))
@@ -238,6 +246,23 @@ let kind_to_constr = function
 
 let hcons_inplace f a = Array.iteri (fun i x -> Array.unsafe_set a i (f x)) a
 
+let partial_array_eqeq a1 a2 n =
+  Array.length a1 = n
+  &&
+  let rec aux i =
+    (Int.equal i n || (a1.(i) == a2.(i) && aux (i+1)))
+  in
+  aux 0
+
+let lookup_partial_app tbl f args n =
+  assert (n > 1); (* or >= ?? *)
+  let hash = hash_app f args n in
+  Tbl.raw_find tbl hash (fun c ->
+      match c.kind with
+      | App (f',args') ->
+        f == f' && partial_array_eqeq args' args n
+      | _ -> false)
+
 let of_kind_nohashcons = function
   | App (c, [||]) -> c
   | kind ->
~~~

**Comment by ppedrot** [comment]
Some quick stats show that this cache is very rarely hit, so not worth it.

**Pull request closed without merging**